### PR TITLE
Shader-based texture AA

### PR DESCRIPTION
**Pull request by temx**
This adds shader-based texture supersampling. This can be used for anti-aliased nearest neighborhood filtered textures:

It also reduces shimmering:
unfiltered             |  supersampled
:-:|:-:
  |  

This is cheaper than per sample shading, and cheaper and higher quality than ordered grid supersampling.

Known limitations:
* cutout alpha textures are not supersampled (would look worse as implemented)
* possible color bleeds since there's no centroid sampling (I have not seen this happen)

### CODE REVIEW

**Comment by tyfighter** [comment]
+1 from me :) as I started looking at exactly the same thing on Wednesday. It looks great!

**Comment by NightFright2k19** [comment]
Looks really cool. Doesn't make graphics look washed out while also preserving the vanilla-ish software renderer visuals.

**Comment by andrei-drexler** [comment]
Looks good and runs well, just a couple of notes:
- unlit water and external fullbrights don't seem to be supersampled, is this intentional? Water UVs are animated, so it might be hard to notice in-game, but it would show up in comparison screenshots.
- `classic antialiased` is too long - maybe `classic 4x`, `classic+AA`, or even just `classic+`?

PS it's possible to achieve virtually perfect antialiasing (instead of just 4x) with a single bilinear sample by tweaking the UVs ([shadertoy demo](https://www.shadertoy.com/view/st23zy)), and distant shimmering could also be fixed by using linear filtering for minification (which looks very similar to the 4 tap approach). In practice this would mean using linear filtering for world textures and nearest filtering for everything else, which would probably make the code a bit messier. It might still be interesting to investigate this approach in the future, though.

**Comment by tyfighter** [comment]
@andrei-drexler - could you do something like this in TexMgr_SetFilterModes to detect world textures? I think this is an excellent idea over supersampling.
```	
	else if (glt->flags & TEXPREF_MIPMAP)
	{
		if (glt->flags & TEXPREF_BINDLESS)
		{	// world texture
			glTexParameterf(glt->target, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
			glTexParameterf(glt->target, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
		}
		else
		{
			glTexParameterf(glt->target, GL_TEXTURE_MAG_FILTER, glmodes[glmode_idx].magfilter);
			glTexParameterf(glt->target, GL_TEXTURE_MIN_FILTER, glmodes[glmode_idx].minfilter);
		}
		glTexParameterf(glt->target, GL_TEXTURE_MAX_ANISOTROPY_EXT, gl_texture_anisotropy.value);
	}
```

**Comment by temx** [comment]
> ([shadertoy demo](https://www.shadertoy.com/view/st23zy))

This is great, I went ahead and implemented it. The problem is that it blurs a little too much in my opinion, especially when minificating:

(4xRGSS)

(Bilinear 16xAF -1 bias)


(4xRGSS)

(Bilinear 16xAF -1 bias)


AF, negative bias and even supersampling don't seem to help. So I tried falling back to nearest 4xRGSS when minificating (this does require both nearest and linear sampling depending on the scale). Magnification does look better on high contrast textures:




But I started having second thoughts for more "normal" textures, where I think it smooths too much:
(4xRGSS)

(Analytic close, 4xRGSS nearest far)

(this one also has the problem of a visible seam when it switches)

In short, I like it more when it looks worse (this is a matter of taste, or lack thereof, of course), so I kept the original approach. 

I pushed a new version that goes ahead and adds support for it for all 3D shaders, incl. unlit water, alias models, and skies. Fullbrights get filtered too. I re-enabled it for cutout alpha, with a lower threshold so it looks better. Changed the menu to Classic+.

The way it's done now, it's easy to add more versions, such as the softer, more accurate one.

**Comment by tyfighter** [comment]
This really is great 👍 the only jaggies are at geometry edges, and at high enough resolutions are barely noticeable.

**Comment by andrei-drexler** [comment]
@tyfighter: Using the bindless flag would work - it would also catch a few builtin textures, but those can be distinguished using the 'persist' flag (plus they're 2x2, solid color, so the filtering mode doesn't really matter anyway). It's definitely one of those things that would warrant a comment explaining what's going on, though.

@temx: Nice work! I only did some very quick testing, so I don't have comparison shots, but looking at yours I agree, the supersampled version does seem like the better choice - it's slightly sharper, and the code is cleaner, too. One final concern, though - I noticed that in the latest code supersampling is a separate shader permutation axis, whereas in the previous version it was grouped with the dithering. On the one hand, it is pretty neat to be able to use this feature with r_softemu, but the number of shaders is getting quite high. Testing this on an old Surface 3 tablet, shader compilation time has gone up from ~1 second to ~2 - not tragic, but not great, either (I guess Intel drivers didn't have a shader cache at that time). At some point I'd like to implement weighted blended OIT, and then the number of shaders will double once more (since transparent versions will need to use 2 render targets). How much of a hassle would it be to go back to the previous permutation setup?

**Comment by temx** [comment]
Would it be OK to compile on demand instead? The current settings at startup, and then any new permutation the first time it's used during the session.

**Comment by andrei-drexler** [comment]
Hmm... that would indeed reduce startup time, but it seems like we'd be fighting complexity with even more complexity.

**Comment by tyfighter** [comment]
@andrei-drexler - The shader combinatorics problem is just the nature of GPU work until true virtual functions are supported. Plus, these shaders aren't changing dynamically, and are only a one time startup cost as they'll hit the driver's shader cache. Another possibility is using [GL_ARB_gl_spirv](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_gl_spirv.txt) and use glSpecializeShaderARB so that the driver can handle all the variants for you.

**Comment by temx** [comment]
Sorry to insist but I gave the "on demand" idea a shot and it ended up costing 2 statements and zero extra state, and makes it easy for future additions to have no additional cost (see added commit). If you think it's going to be a maintenance problem, I can move it to be another softemu value as originally.

**Comment by tyfighter** [comment]
@temx - Just wanted to point out that there is one more issue with this. The defaults for GL_TEXTURE_WRAP_(S|T) is GL_REPEAT. When doing the 4 sample taps, you can read from outside the texture bounds. This can sometimes create tiny artifacts at geometry edges when the texel coordinate sampled modulo texture size on the opposite side of the texture does not fit with the local texels. Thankfully in Quake, there are a lot of repeating textures, so this isn't noticeable, but I've now seen it in a few. It might be worth trying
```
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
```
or maybe even

```
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_BORDER);
glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_BORDER);
float color[] = { 0.0f, 0.0f, 0.0f, 0.0f };
glTexParameterfv(GL_TEXTURE_2D, GL_TEXTURE_BORDER_COLOR, color);
```

**Comment by temx** [comment]
@tyfighter I know, I alluded to it when I said "possible color bleeds". Unfortunately at the moment textures have to tile, the UVs are more often than not outside the (0, 1) range. If you clamp to border you get this:

**Comment by tyfighter** [comment]
Darn :( oh well. Thankfully, they really are barely noticeable. I guess to fix it the shader would need to know whether or not the texture repeats to avoid it, and that information just doesn't exist.

**Comment by andrei-drexler** [comment]
@tyfighter

> The shader combinatorics problem is just the nature of GPU work until true virtual functions are supported.

The fact that a problem is very common doesn't mean that we should just accept it, on the contrary, we should be mindful of it and try to keep it in check every step of the way. For a game that looks like Quake, though, I don't think this problem should exist at all.

> Plus, these shaders aren't changing dynamically, and are only a one time startup cost as they'll hit the driver's shader cache. Another possibility is using [GL_ARB_gl_spirv](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_gl_spirv.txt) and use glSpecializeShaderARB so that the driver can handle all the variants for you.

Unfortunately the system I mentioned (Surface 3) doesn't have either SPIR-V support or apparently even a driver-side shader cache. And this engine runs on hardware even older than that, the only hard requirement being OpenGL 4.3. My concern is not just about the extra shaders, though, it's also about the added complexity, which I'm trying to minimize.

---

@temx:

> Sorry to insist but I gave the "on demand" idea a shot and it ended up costing 2 statements and zero extra state, and makes it easy for future additions to have no additional cost (see added commit). If you think it's going to be a maintenance problem, I can move it to be another softemu value as originally.
>
> Edit: first attempt was broken :( I guess that's a good argument against.

In this case it's also about trying to avoid runtime shader compilation (which I think is a slippery slope), in addition to the maintenance problem that already seems to have made its presence felt a bit. Even small increases in code complexity can have significant repercussions over time, especially on a hobby project where a lot of the code is written late at night, when mistakes are more likely to be made. That's one reason I want to keep things as simple as possible.

I've been thinking a lot about what to do here. Before noticing the (admittedly very subtle) occasional edge artifacts, I thought this was a viable replacement for per-sample shading. I wasn't too happy about the increased shader/code complexity, but it seemed like a reasonable trade-off for a solution that is faster than the alternative and works even in non-MSAA mode. Given those tiny imperfections, though, and the original goal of improving image quality, I think it makes a lot of sense to also consider per-sample shading. It's just a few lines of self-contained code, it doesn't have issues at geometry boundaries, and it also allows the user to balance image quality and performance by varying the AA level. Once per-sample shading is added, though, texture super-sampling becomes redundant in my opinion. There is still a small niche where using it might make sense, but I no longer think it's worth modifying all of the shaders for it and potentially confusing players with two separate options that have roughly the same effect.

I hope this sounds like a fair and reasonable assessment, and for what it's worth, I try to apply the same principles to my own work - I have a folder full of experiments, many of which haven't made the cut. I realize how demotivating this might be, especially considering the extra work you've done at my request, and I apologize for not seeing the bigger picture more clearly from the beginning and for wasting your time. I really hope I'm not burning a bridge here, and that this doesn't negate your interest in future contributions.

**Comment by temx** [comment]
No problem! I was also not happy with how large of a change it turned out to be in the end, the only advantage being faster performance against image quality (not just borders, but more prominently cutout alpha).
